File: common/src/main/java/dev/thomasqtruong/veryscuffedcobblemonbreeding/util/ItemBuilder.java
package dev.thomasqtruong.veryscuffedcobblemonbreeding.util;

import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtList;
import net.minecraft.nbt.NbtString;
import net.minecraft.text.Style;
import net.minecraft.text.Text;
import net.minecraft.text.Texts;

public class ItemBuilder {
    ItemStack stack = null;

    public ItemBuilder(Item item) {
        this.stack = new ItemStack(item);
    }
    public ItemBuilder(ItemStack item) {
        this.stack = item;
    }

    public ItemBuilder addLore(Text[] lore) {
        NbtCompound nbt = this.stack.getOrCreateNbt();
        NbtCompound displayNbt = this.stack.getOrCreateSubNbt("display");
        NbtList nbtLore = new NbtList();

        for (Text text : lore) {
            Text line = Texts.join(text.getWithStyle(Style.EMPTY.withItalic(false)), Text.of(""));
            nbtLore.add(NbtString.of(Text.Serializer.toJson(line)));
        }

        displayNbt.put("Lore", nbtLore);
        nbt.put("display", displayNbt);
        this.stack.setNbt(nbt);
        return this;
    }

    public ItemBuilder hideAdditional() {
        this.stack.addHideFlag(ItemStack.TooltipSection.ADDITIONAL);
        return this;
    }
    public ItemBuilder setCustomName(Text customName) {
        Text pokemonName = Texts.join(customName.getWithStyle(Style.EMPTY.withItalic(false)), Text.of(""));
        this.stack.setCustomName(pokemonName);
        return this;
    }

    public ItemStack build() {
        return this.stack;
    }
}

File: common/src/main/kotlin/dev/thomasqtruong/veryscuffedcobblemonbreeding/util/ItemBuilder.kt
package dev.thomasqtruong.veryscuffedcobblemonbreeding.util

import net.minecraft.core.component.DataComponents
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.chat.Component
import net.minecraft.util.Unit
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.component.CustomData
import net.minecraft.world.item.component.CustomModelData
import net.minecraft.world.item.component.ItemLore
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream


class ItemBuilder {
    var stack: ItemStack? = null

    constructor(item: Item?) {
        this.stack = ItemStack(item)
    }

    constructor(item: ItemStack?) {
        this.stack = item
    }

    fun addLore(newLore: Array<Component>?): ItemBuilder {
        var itemLore = stack!!.get(DataComponents.LORE)

        if (itemLore == null) {
            itemLore = ItemLore(listOf())
        }

        var list = Stream.concat(itemLore.lines().stream(), Arrays.stream(newLore)).collect(Collectors.toList())
        // Go through every Component, and get the current style and set Italic to false
        list = list.stream().map { component: Component -> component.copy().withStyle(component.style.withItalic(false)) }.collect(Collectors.toList()) as List<Component>?

        itemLore = ItemLore(list)
        stack!!.set(DataComponents.LORE, itemLore)
        return this
    }

    fun setCustomModel(customModelVal: Int): ItemBuilder {
        this.stack!!.set(DataComponents.CUSTOM_MODEL_DATA, CustomModelData(customModelVal))
        return this
    }


    fun setCustomData(data: CustomData): ItemBuilder {
        // If the stack has custom data, get it and append the new data to it, otherwise create a new custom data with the new data
        val customData = if (stack!!.has(DataComponents.CUSTOM_DATA)) stack!!.get(DataComponents.CUSTOM_DATA) else CustomData.of(CompoundTag())

        val newTag = customData!!.copyTag()
        val tag = data.copyTag()

        tag.allKeys.forEach { key -> newTag.put(key, tag.get(key)) }

        stack!!.set(DataComponents.CUSTOM_DATA, CustomData.of(newTag))
        return this
    }

    fun setAmount(amount: Int): ItemBuilder {
        this.stack!!.setCount(amount);
        return this;
    }

    fun hideAdditional(): ItemBuilder {
//        this.stack.addHideFlag(ItemStack.TooltipSection.ADDITIONAL);
        stack!!.set(DataComponents.HIDE_ADDITIONAL_TOOLTIP, Unit.INSTANCE)
        return this
    }

    fun setCustomName(customName: Component?): ItemBuilder {
//        Text pokemonName = Texts.join(customName.getWithStyle(Style.EMPTY.withItalic(false)), Component.literal(""));
//        this.stack.setCustomName(pokemonName);
        stack!!.set(DataComponents.CUSTOM_NAME, customName)

        return this
    }

    fun build(): ItemStack {
        return this.stack!!
    }
}